### PR TITLE
Bring 1107-intermediatestatus to main branch for the feature

### PR DESCRIPTION
**Pull request by rjzondervan**

### CODE REVIEW

File: api/src/Service/ResponseService.php [hunk 1]
```diff
@@ -377,7 +432,36 @@ private function renderValues(ObjectEntity $result, ?array $fields, ?array $exte
...
+                    }
+                    foreach ($objects as $object) {
+                        if ($acceptType === 'jsonld') {
+                            $response[$attribute->getName()][] = [
+                                '@id' => $object->getSelf() ?? '/api/'.($object->getEntity()->getRoute() ?? $object->getEntity()->getName()).'/'.$object->getId(),
+                            ];
+                        } else {
+                            $response[$attribute->getName()][] = $object->getSelf() ?? '/api/'.($object->getEntity()->getRoute() ?? $object->getEntity()->getName()).'/'.$object->getId();
+                        }
+                    }
```
Comment (smisidjan) [inline api/src/Service/ResponseService.php:464]: This are a lot of nested levels

**Review by smisidjan — CHANGES_REQUESTED** [review]

File: api/src/Service/ResponseService.php (hunk 1 above)
Comment (rjzondervan) [reply api/src/Service/ResponseService.php:464]: @WilcoLouwerse Is this something you refactor quickly, or do I have to look into it?

File: api/src/Entity/Action.php [hunk 2]
```diff
@@ -2,49 +2,139 @@
...
     /**
-     * @ORM\Column(type="array", nullable=true)
+     * @var array The event names the action should trigger
+     *
+     * @Groups({"read","read_secure","write"})
+     *
+     * @ORM\Column(type="simple_array", nullable=true)
      */
-    private $configuration = [];
+    private array $throws = [];
```
Comment (WilcoLouwerse) [inline api/src/Entity/Action.php:91]: ?array want nullable=true ?

File: api/src/Entity/Action.php [hunk 3]
```diff
@@ -2,49 +2,139 @@
...
     /**
-     * @ORM\Column(type="string", length=255)
+     * @var array The conditions that the data object should match for the action to be triggered
+     *
+     * @Groups({"read","read_secure","write"})
+     *
+     * @ORM\Column(type="json", nullable=true)
      */
-    private $type;
+    private array $conditions = [];
```
Comment (WilcoLouwerse) [inline api/src/Entity/Action.php:100]: ?array want nullable=true ?

File: api/src/Entity/Action.php [hunk 4]
```diff
@@ -2,49 +2,139 @@
...
+     * @ORM\Column(type="boolean")
+     */
+    private bool $async = false;
+
+    /**
+     * @var array The configuration of the action
+     * @Groups({"read","read_secure","write"})
+     * @ORM\Column(type="array", nullable=true)
+     */
+    private array $configuration = [];
```
Comment (WilcoLouwerse) [inline api/src/Entity/Action.php:135]: ?array want nullable=true ?

File: api/src/Entity/Synchronization.php [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+     * @ORM\GeneratedValue(strategy="CUSTOM")
+     * @ORM\CustomIdGenerator(class="Ramsey\Uuid\Doctrine\UuidGenerator")
+     */
+    private UuidInterface $id;
+
+    /**
+     * @var Entity The entity of this resource
+     *
+     * @Groups({"read","write"})
+     * @ORM\OneToOne(targetEntity=Entity::class, cascade={"persist", "remove"})
```
Comment (WilcoLouwerse) [inline api/src/Entity/Synchronization.php:66]: Should be manyToOne and pls no cascading here!

Comment (WilcoLouwerse) [reply api/src/Entity/Synchronization.php:66]: This branch is and should be behind here i think?

File: api/src/Entity/Synchronization.php [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+     * @ORM\JoinColumn(nullable=false)
+     */
+    private entity $entity;
+
+    /**
+     * @var ?ObjectEntity The object of this resource
+     *
+     * @Groups({"read","write"})
+     * @ORM\OneToOne(targetEntity=ObjectEntity::class, cascade={"persist", "remove"})
```
Comment (WilcoLouwerse) [inline api/src/Entity/Synchronization.php:75]: Same as for $entity above^

**Review by WilcoLouwerse — APPROVED** [review]
Read my comments please before merging ;)
And make sure synchronization.php is updated, this is an old version of the file/entity

File: api/src/Entity/Action.php (hunk 2 above)
Comment (rjzondervan) [reply api/src/Entity/Action.php:91]: It's defaulted by = [];

File: api/src/Entity/Action.php (hunk 3 above)
Comment (rjzondervan) [reply api/src/Entity/Action.php:100]: It's defaulted by = [];

File: api/src/Entity/Action.php (hunk 4 above)
Comment (rjzondervan) [reply api/src/Entity/Action.php:135]: It's defaulted by = [];

File: api/src/Service/ResponseService.php (hunk 1 above)
Comment (WilcoLouwerse) [reply api/src/Service/ResponseService.php]: This is something that still contains a bug i need to fix :)
after that we can think about a refactor

File: api/src/Entity/Action.php (hunk 4 above)
Comment (WilcoLouwerse) [reply api/src/Entity/Action.php:135]: see comment on $throws

File: api/src/Service/ResponseService.php (hunk 1 above)
Comment (WilcoLouwerse) [reply api/src/Service/ResponseService.php]: Bug seems to only happen in one specific online installation. not local.
Will look to improve/cleanup this code tomorrow.

File: api/src/Service/ResponseService.php [hunk 7]
```diff
@@ -509,6 +472,82 @@ private function renderValues(ObjectEntity $result, ?array $fields, ?array $exte
...
+            if (array_key_exists('all', $extend)) {
+                $subExtend = $extend;
+            } elseif (array_key_exists($attribute->getName(), $extend)) {
+                if (is_array($extend[$attribute->getName()])) {
+                    $subExtend = $extend[$attribute->getName()];
+                } elseif ($extend[$attribute->getName()] == false) {
+                    return [];
+                }
+            }
+        }
```
Comment (rjzondervan) [inline api/src/Service/ResponseService.php:550]: You've still got a lot of nesting levels here, maybe it's better to have three times elseif than an if in an elseif in an if.

File: api/src/Service/ResponseService.php [hunk 8]
```diff
@@ -509,6 +472,82 @@ private function renderValues(ObjectEntity $result, ?array $fields, ?array $exte
...
+                if ($acceptType === 'jsonld') {
+                    $response[$attribute->getName()][] = [
+                        '@id' => $object->getSelf() ?? '/api/'.($object->getEntity()->getRoute() ?? $object->getEntity()->getName()).'/'.$object->getId(),
+                    ];
+                } else {
+                    $response[$attribute->getName()][] = $object->getSelf() ?? '/api/'.($object->getEntity()->getRoute() ?? $object->getEntity()->getName()).'/'.$object->getId();
+                }
+            }
+            return false;
+        }
```
Comment (rjzondervan) [inline api/src/Service/ResponseService.php:523]: Same as previous comment, still a large number of nesting levels, making the code complex and not necessarily readable

**Review by rjzondervan — COMMENTED** [review]
Still some nesting problems

**Review by smisidjan — APPROVED** [review]

**Pull request merged**